pkg/clientcert/cert_controller.go

```diff
@@ -0,0 +1,344 @@
+	csrName string
+
+	// keyData is the private key data used to created a csr
+	// csrName and keyData store the internal state of the controller. They are set after controller creates a new csr
+	// and cleared once the csr is approved and processed by controller. There are 4 combination of their values:
+	//   1. csrName empty, keyData empty: means we aren't trying to create a new client cert, our current one is valid
+	//   2. csrName set, keyData empty: there was bug
+	//   3. csrName set, keyData set: we are waiting for a new cert to be signed.
+	//   4. csrName empty, keydata set: the CSR failed to create, this shouldn't happen, it's a bug.
+	keyData []byte
+}
+
+// NewClientCertificateController return an instance of clientCertificateController
+func NewClientCertificateController(
+	clientCertSecretNamespace string,
+	clientCertSecretName string,
+	hubCSRInformer certificatesinformers.CertificateSigningRequestInformer,
+	hubCSRClient csrclient.CertificateSigningRequestInterface,
+	spokeSecretInformer corev1informers.SecretInformer,
+	spokeCoreClient corev1client.CoreV1Interface,
+	additonalSecretData map[string][]byte,
+	csrObjectMeta metav1.ObjectMeta,
+	csrSubject *pkix.Name,
+	csrSignerName string,
+	csrFilterFunc factory.EventFilterFunc,
+	recorder events.Recorder,
+	controllerName string,
+) factory.Controller {
+	c := clientCertificateController{
+		clientCertSecretNamespace: clientCertSecretNamespace,
+		clientCertSecretName:      clientCertSecretName,
+		hubCSRLister:              hubCSRInformer.Lister(),
+		hubCSRClient:              hubCSRClient,
+		spokeCoreClient:           spokeCoreClient,
+		additonalSecretData:       additonalSecretData,
+		csrObjectMeta:             csrObjectMeta,
+		csrSubject:                csrSubject,
+		csrSignerName:             csrSignerName,
+		controllerName:            controllerName,
+	}
+
+	return factory.New().
+		WithFilteredEventsInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, func(obj interface{}) bool {
+			accessor, err := meta.Accessor(obj)
+			if err != nil {
+				return false
+			}
+			// only enqueue a specific secret
+			if accessor.GetNamespace() == clientCertSecretNamespace && accessor.GetName() == clientCertSecretName {
+				return true
+			}
+			return false
+		}, spokeSecretInformer.Informer()).
+		WithFilteredEventsInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, csrFilterFunc, hubCSRInformer.Informer()).
+		WithSync(c.sync).
+		ResyncEvery(ControllerResyncInterval).
+		ToController(controllerName, recorder)
+}
+
+func (c *clientCertificateController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	// get secret containing client certificate
+	secret, err := c.spokeCoreClient.Secrets(c.clientCertSecretNamespace).Get(ctx, c.clientCertSecretName, metav1.GetOptions{})
```

Comment:
followup: as you start running more of these, you may end up wanting to use listers.